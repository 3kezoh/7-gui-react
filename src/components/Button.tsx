import { cn } from "../utils";

type ButtonProps = React.ComponentPropsWithoutRef<"button">;

export function Button(props: ButtonProps) {
  const { className, disabled, ...rest } = props;

  const buttonClassName = cn(
    "border border-black px-2",
    className,
    {
      "transition-colors hover:text-white hover:bg-black": !disabled,
    },
    {
      "cursor-not-allowed bg-gray-200 text-gray-400 border-gray-400": disabled,
    },
  );

  return (
    <button
      className={buttonClassName}
      disabled={disabled}
      type="button"
      {...rest}
    />
  );
}
