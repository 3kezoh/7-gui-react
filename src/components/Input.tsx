import { cn } from "../utils";

type InputProps = React.ComponentPropsWithoutRef<"input">;

export function Input(props: InputProps) {
  const { className, name, ...rest } = props;
  const inputClassName = cn("border border-black w-full max-w-xs", className);

  return (
    <input
      className={inputClassName}
      type="text"
      id={name}
      name={name}
      {...rest}
    />
  );
}
