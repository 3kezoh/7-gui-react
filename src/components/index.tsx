export * from "./Counter";
export * from "./FlightBooker";
export * from "./TemperatureConverter";
