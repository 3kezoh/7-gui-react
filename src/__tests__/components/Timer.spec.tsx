import { act, fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Timer } from "../../components";

beforeEach(() => {
  vi.useFakeTimers();
});

afterEach(() => {
  vi.runOnlyPendingTimers();
  vi.useRealTimers();
});

describe("Timer", () => {
  it("should be in the document", () => {
    render(<Timer />);

    const timer = screen.getByTestId("timer");

    expect(timer).toBeInTheDocument();
  });

  describe("As time goes by", () => {
    it("should display the elapsed time", () => {
      render(<Timer />);

      act(() => vi.advanceTimersByTime(100));

      const timeElement = screen.getByText("0.1s");

      expect(timeElement).toBeInTheDocument();
    });

    it("should have a progress bar that fills up", () => {
      render(<Timer />);

      act(() => vi.advanceTimersByTime(500));

      const progressElement = screen.getByRole<HTMLInputElement>("progressbar");

      expect(progressElement).toHaveValue(500);
    });
  });

  describe("After one second", () => {
    it("should display 1s", async () => {
      render(<Timer />);

      act(() => vi.advanceTimersByTime(1000));

      const timeElement = screen.getByText("1s");

      expect(timeElement).toBeInTheDocument();
    });

    it("should have the progress bar filled to the maximum", async () => {
      render(<Timer />);

      act(() => vi.advanceTimersByTime(1000));

      const progressElement = screen.getByRole<HTMLInputElement>("progressbar");

      expect(progressElement).toHaveValue(progressElement.max);
    });
  });

  describe("Using the duration slider", () => {
    it("should update the progress bar maximum", async () => {
      render(<Timer />);

      const sliderElement = screen.getByRole("slider");

      fireEvent.change(sliderElement, { target: { value: 2000 } });

      const progressElement = screen.getByRole<HTMLInputElement>("progressbar");

      expect(progressElement).toHaveAttribute("max", "2000");
    });

    it("should update the timer duration", async () => {
      render(<Timer />);

      const sliderElement = screen.getByRole("slider");

      fireEvent.change(sliderElement, { target: { value: 2000 } });

      act(() => vi.advanceTimersByTime(2000));

      const timeElement = screen.getByText("2s");

      expect(timeElement).toBeInTheDocument();
    });

    it("should make the timer pick up where it left off", async () => {
      render(<Timer />);

      act(() => vi.advanceTimersByTime(1000));

      const progressElement = screen.getByRole<HTMLInputElement>("progressbar");

      expect(progressElement).toHaveValue(progressElement.max);

      const sliderElement = screen.getByRole("slider");

      fireEvent.change(sliderElement, { target: { value: 2000 } });

      act(() => vi.advanceTimersByTime(3000));

      const timeElement = screen.getByText("4s");

      expect(timeElement).toBeInTheDocument();
    });
  });

  describe("A click on the reset button", () => {
    it("should reset the displayed time", async () => {
      const user = userEvent.setup({ delay: null });

      render(<Timer />);

      const resetElement = screen.getByRole("button", { name: /reset/i });

      await user.click(resetElement);

      const timeElement = screen.getByText("0s");

      expect(timeElement).toBeInTheDocument();
    });

    it("should reset the progress bar", async () => {
      const user = userEvent.setup({ delay: null });

      render(<Timer />);

      const resetElement = screen.getByRole("button", { name: /reset/i });

      await user.click(resetElement);

      const progressElement = screen.getByRole("progressbar");

      expect(progressElement).toHaveValue(0);
    });

    it("should not reset the duration", async () => {
      const user = userEvent.setup({ delay: null });

      render(<Timer />);

      const resetElement = screen.getByRole("button", { name: /reset/i });

      await user.click(resetElement);

      const sliderElement = screen.getByRole("slider");

      expect(sliderElement).toHaveValue("1000");
    });
  });

  it("should match the snapshot", () => {
    const { asFragment } = render(<Timer />);

    expect(asFragment()).toMatchSnapshot();
  });
});
