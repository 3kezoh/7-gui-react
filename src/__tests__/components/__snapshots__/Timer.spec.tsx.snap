// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Timer > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid gap-4 p-4 border border-black w-fit"
    data-testid="timer"
  >
    <div
      class="flex gap-2"
    >
      <label
        for="time"
      >
        Elapsed Time:
      </label>
      <progress
        id="time"
        max="1000"
        value="0"
      />
    </div>
    <div>
      0s
    </div>
    <div
      class="flex gap-2"
    >
      <label
        for="duration"
      >
        Duration:
      </label>
      <input
        class="border border-black w-full max-w-xs"
        id="duration"
        max="10000"
        min="0"
        name="duration"
        step="100"
        type="range"
        value="1000"
      />
    </div>
    <button
      class="border border-black px-2"
      type="button"
    >
      Reset
    </button>
  </div>
</DocumentFragment>
`;
